Indicators/McGinleyDynamic.cs

```diff
@@ -0,0 +1,92 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// Represents the McGinley Dynamic (MGD)
+    /// </summary>
+    public class McGinleyDynamic : WindowIndicator<IndicatorDataPoint>, IIndicatorWarmUpPeriodProvider
+    {
+        /// <summary>
+        /// A rolling sum for computing the average for the given period
+        /// </summary>
+        private IndicatorBase<IndicatorDataPoint> _rollingSum { get; }
+
+        private readonly int _period;
+
+        /// <summary>
+        /// Gets a flag indicating when this indicator is ready and fully initialized
+        /// </summary>
+        public override bool IsReady => Samples >= _period;
+
+        /// <summary>
+        /// Resets this indicator to its initial state
+        /// </summary>
+        public override void Reset()
+        {
+            _rollingSum.Reset();
+            base.Reset();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the McGinleyDynamic class with the specified name and period
+        /// </summary>
+        /// <param name="name">The name of this indicator</param>
+        /// <param name="period">The period of the McGinley Dynamic</param>
+        public McGinleyDynamic(string name, int period)
+            : base(name, period)
+        {
+            _period = period;
+            _rollingSum = new Sum(name + "_Sum", period);
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the McGinleyDynamic class with the default name and period
+        /// </summary>
+        /// <param name="period">The period of the McGinley Dynamic</param>
+        public McGinleyDynamic(int period)
+            : this($"MGD({period})", period)
+        {
+        }
+
+        /// <summary>
+        /// Computes the next value for this indicator from the given state.
+        /// </summary>
+        /// <param name="window">The window of data held in this indicator</param>
+        /// <param name="input">The input value to this indicator on this time step</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(IReadOnlyWindow<IndicatorDataPoint> window, IndicatorDataPoint input)
+        {
+            _rollingSum.Update(input.Time, input.Value);
+            if (Samples < _period) {
+                return 0;
+            } else if (Samples == _period) {
+                return _rollingSum.Current.Value / _period;
+            } else {
+                if (Current.Value == 0)
+                    return Current.Value;
+                return Current.Value + (input.Value - Current.Value) / (_period * (decimal)Math.Pow((double)(input.Value / Current.Value), 4.0));
```

Comment:
We should probably add a protection against `/ 0` or infinity explosions

Reply:
nit - missing `{}` too
Should rebase after https://github.com/QuantConnect/Lean/pull/8177 👍